### PR TITLE
GH-231 Added artifactory_virtual_conan_repository resource

### LINKED ISSUE #231: add artifactory_virtual_conan_repository
## Version
2.6.8

The following resource recently starting giving deprecation warnings in the logs

```hcl
resource "artifactory_virtual_repository" "conan" {
  key          = "conan"
  package_type = "conan"
  repositories = [
    "conan-local",
    "conan-remote"
  ]
  default_deployment_repo = "conan-local"
  notes                   = "managed by terraform"
}
```

```bash
Warning: Deprecated Resource

  with artifactory_virtual_repository.conan,
  on repository-conan.tf line 18, in resource "artifactory_virtual_repository" "conan":
  18: resource "artifactory_virtual_repository" "conan" {

This resource is deprecated and you should use repo type specific resources
(such as artifactory_virtual_maven_repository) in the future
```

I _assume_ the replacement should be a resource type of `artifactory_virtual_conan_repository`. That resource type doesn't exist. 

```
│ Error: Invalid resource type
│ 
│   on repository-conan.tf line 18, in resource "artifactory_virtual_conan_repository" "conan":
│   18: resource "artifactory_virtual_conan_repository" "conan" {
│ 
│ The provider jfrog/artifactory does not support resource type "artifactory_virtual_conan_repository".
```

The only valid resource types according to the documentation are

1. `artifactory_virtual_maven_repository`
2. `artifactory_virtual_go_repository`

There is no `artifactory_virtual_conan_repository`

**What should be done about this deprecation? Should I ignore it? Or are there plans for a `artifactory_virtual_conan_repository` resource?**

### PR DESCRIPTION
**Pull request by maheshjfrog**
Fixes #231

### CODE REVIEW

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+type ConanVirtualRepositoryParams struct {
+	VirtualRepositoryBaseParams
+	VirtualRetrievalCachePeriodSecs int `hcl:"virtual_retrieval_cache_period_seconds" json:"virtualRetrievalCachePeriodSecs,omitempty"`
+}
+
+func resourceArtifactoryConanVirtualRepository() *schema.Resource {
+	var conanVirtualSchema = mergeSchema(baseVirtualRepoSchema, map[string]*schema.Schema{
+		"virtual_retrieval_cache_period_seconds": {
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_virtual_conan_repository.go:15]: Do we need the prefix `virtual_` here? The resource already contains the word `virtual`

File: pkg/artifactory/resource_artifactory_virtual_repository_test.go [hunk 2]
```diff
@@ -144,6 +144,38 @@ func TestAccVirtualGoRepository_basic(t *testing.T) {
...
+		CheckDestroy:      verifyDeleted(fqrn, testCheckRepo),
+		ProviderFactories: testAccProviders,
+
+		Steps: []resource.TestStep{
+			{
+				Config: virtualRepositoryBasic,
+				// we check to make sure some of the base params are picked up
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "key", name),
+					resource.TestCheckResourceAttr(fqrn, "package_type", "conan"),
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_virtual_repository_test.go:172]: Please add check for the new attribute.

**Review by alexhung — CHANGES_REQUESTED** [review]

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: "This value refers to the number of seconds to cache metadata files before checking for newer versions on aggregated repositories. A value of 0 indicates no caching.",
+			DefaultFunc: func() (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go:20]: Looks like "Computed: true" setting ... blocks us from using "Default: 7200". We get "Default must be nil if computed" as response.
Computed setting does't make sense here. Will remove and use Default: 7200.

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package artifactory
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+type ConanVirtualRepositoryParams struct {
+	VirtualRepositoryBaseParams
+	VirtualRetrievalCachePeriodSecs int `hcl:"virtual_retrieval_cache_period_seconds" json:"virtualRetrievalCachePeriodSecs,omitempty"`
```
[the remark this answers is not in the document]
Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go]: removed omitempty setting.

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go (hunk 1 above)
Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go]: made it retrieval_cache_period_seconds now. ie: removed virtual in name as suggested.

File: pkg/artifactory/resource_artifactory_virtual_repository_test.go (hunk 2 above)
Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_repository_test.go:172]: added.

File: docs/resources/artifactory_virtual_conan_repository.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# Artifactory Virtual Conan Repository Resource
+
+Provides an Artifactory virtual repository resource, but with specific conan features. This should be preferred over the original
+one-size-fits-all `artifactory_virtual_repository`. 
+
+## Example Usage
+
+```hcl
+resource "artifactory_virtual_conan_repository" "foo-conan" {
```
Comment (chb0github) [inline docs/resources/artifactory_virtual_conan_repository.md:9]: You see, I only include the things that are unique to conan - everything here seems to be just like the base repo

File: docs/resources/artifactory_virtual_conan_repository.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+* `key` - (Required)
+* `repositories` - (Required, but may be empty)
+* `description` - (Optional)
+* `notes` - (Optional)
+* `includes_pattern` - (Optional)
+* `excludes_pattern` - (Optional)
+* `repo_layout_ref` - (Optional)
+* `artifactory_requests_can_retrieve_remote_artifacts` - (Optional)
+* `key_pair` - (Optional) - Key pair to use for... well, I'm not sure. Maybe ssh auth to remote repo?
```
Comment (chb0github) [inline docs/resources/artifactory_virtual_conan_repository.md:32]: It'd be great if someone could figure it out

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+			VirtualRepositoryBaseParams: VirtualRepositoryBaseParams{
+				Rclass:      "virtual",
+				PackageType: "conan",
+			},
+		}
+	})
+
+}
+
+func unpackConanVirtualRepository(s *schema.ResourceData) (interface{}, string, error) {
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_virtual_conan_repository.go:35]: Move this inside the function at line 13 to totally hide it

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	d := &ResourceData{s}
+	repo := ConanVirtualRepositoryParams{
+		VirtualRepositoryBaseParams:     unpackBaseVirtRepo(s),
+		VirtualRetrievalCachePeriodSecs: d.getInt("retrieval_cache_period_seconds", true),
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_virtual_conan_repository.go:39]: You're using true - are you sure that's what you want?

Comment (chb0github) [inline pkg/artifactory/resource_artifactory_virtual_conan_repository.go:39]: Also, is this field conan specific?

**Comment by chb0github** [comment]
I don't see anything conan specific here (not suprised) - and, that's fine. If it isn't, then really define a "generic" virtual repo with a parameterized type (like we do with local repos)

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go (hunk 8 above)
Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go:39]: It has to be false. Will change. Thanks.

Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go:39]: This is common. Artifactory UI shows this field in different tab for few package types which confused me. But, Yes, all packages have this setting.

Comment (chb0github) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go:39]: Ok, so if it's common to all virtual repos, then move this code to `unpackBaseVirtualRepo` and then the only difference is that the type is `conan`

This means you should be doing this like you did for local repos. @alexhung , thoughts?

So, basically you have a generic virtual repo, that takes a param of `type` and in this base you pass it `conan` - it also means implementing anything without special fields can be done effortlessly

Comment (alexhung) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go]: Agree! If we have a group of identical virtual repos (except type), then we should create same function like the local one and build once, use many times.

File: docs/resources/artifactory_virtual_conan_repository.md (hunk 6 above)
Comment (maheshjfrog) [reply docs/resources/artifactory_virtual_conan_repository.md:32]: From UI, I see 'Key Pair' shown for maven repo type in advanced tab of virtual repo creation page. But, Not for conan, go, generic, etc.

Comment (alexhung) [reply docs/resources/artifactory_virtual_conan_repository.md:32]: Yeah, there are definitely common attributes but also some very type specific attributes. Probably can't use one generic function for all virtual repos.

Comment (chb0github) [reply docs/resources/artifactory_virtual_conan_repository.md:32]: Not for all of them, but for a lot of them - in fact, it's probably the exact same one as generic

File: pkg/artifactory/repositories.go [hunk 9]
```diff
@@ -700,22 +708,23 @@ func getDefaultDeploymentRepo(d *ResourceData) string {
...
 		IncludesPattern: d.getString("includes_pattern", false),
 		ExcludesPattern: d.getString("excludes_pattern", false),
 		RepoLayoutRef:   d.getString("repo_layout_ref", false),
 		ArtifactoryRequestsCanRetrieveRemoteArtifacts: d.getBool("artifactory_requests_can_retrieve_remote_artifacts", false),
-		Repositories:          d.getList("repositories"),
-		Description:           d.getString("description", false),
-		Notes:                 d.getString("notes", false),
-		DefaultDeploymentRepo: getDefaultDeploymentRepo(d),
+		Repositories:                    d.getList("repositories"),
+		Description:                     d.getString("description", false),
```
Comment (chb0github) [inline pkg/artifactory/repositories.go:724]: `gofmt` ?

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go (hunk 8 above)
Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go]: i mixed up "retrieval_cache_period_seconds" with "artifactoryRequestsCanRetrieveRemoteArtifacts".
UI, shows "retrieval_cache_period_seconds" for some package type and not for others.
So, will make this specific for Conan virtual repo.

**Review by chb0github — APPROVED** [review]
Though not a part of this PR, we can/should now just implement support for all repos that fit this shape. It should be like, a 10 minute job

File: pkg/artifactory/repositories.go (hunk 9 above)
Comment (maheshjfrog) [reply pkg/artifactory/repositories.go:724]: did run gofmt ... will check why the alignment is different.

File: pkg/artifactory/resource_artifactory_virtual_conan_repository.go (hunk 8 above)
Comment (maheshjfrog) [reply pkg/artifactory/resource_artifactory_virtual_conan_repository.go]: Created resourceArtifactoryVirtualGenericRepository function for generic like package type.
And, resourceArtifactoryVirtualRepositoryWithRetrievalCachePeriodSecs function for package types with retrieval_cache_period_seconds setting.

**Review by chb0github — APPROVED** [review]

**Review by alexhung — APPROVED** [review]

**Pull request merged**